# Esercitazione 4 Marzo 2025

## Esercizio 1
- Costruire due array numpy-array 2d di dimensione 5000x5000 i cui valori sono campionati da una distribuzione gaussiana con media 0 e varianza 1.
- Effetuare la somma tra le due matrici con due metodi:

        -  utilizzando le operazioni vettoriali 
        -  considerando due cicli espliciti annidati, su righe e colonne
- Confrontare le prestazioni in termini di tempo di esecuzione.
        


In [9]:
import numpy as np
import time as t

arr1 = np.random.randn(5000, 5000)
arr2 = np.random.randn(5000, 5000)

start = t.time()
arr3 = arr1 + arr2
print("Numpy's sum:", t.time() - start)

start = t.time()
arr4 = np.zeros_like(arr1)
for i in range(5000):
    for j in range(5000):
        arr4[i, j] = arr1[i, j] + arr2[i, j]
print("Python's sum:", t.time() - start)

Numpy's sum: 0.05822277069091797
Python's sum: 14.38627815246582


## Esercizio 2
-      Costruire il numpy array  s  formato da 40 valori equidistanti in  [0,20],
-      considerarne il reshape in un array 2d che ha 4 righe con un numero opportuno di colonne e dire se il reshape crea una vista o una copia

In [12]:
s = np.linspace(0, 20, 40)
b = s.reshape(4, -1)
print(b.base)

[ 0.          0.51282051  1.02564103  1.53846154  2.05128205  2.56410256
  3.07692308  3.58974359  4.1025641   4.61538462  5.12820513  5.64102564
  6.15384615  6.66666667  7.17948718  7.69230769  8.20512821  8.71794872
  9.23076923  9.74358974 10.25641026 10.76923077 11.28205128 11.79487179
 12.30769231 12.82051282 13.33333333 13.84615385 14.35897436 14.87179487
 15.38461538 15.8974359  16.41025641 16.92307692 17.43589744 17.94871795
 18.46153846 18.97435897 19.48717949 20.        ]


## Esercizio 3
-     Costruire il numpy array t formato da 100 valori equidistanti in  [0,4]
-     considerarne il reshape in un array 2d che ha 4 colonne con un numero opportuno di righe

In [13]:
t = np.linspace(0, 4, 100)
b = t.reshape(-1, 4)
print(b.base)

[0.         0.04040404 0.08080808 0.12121212 0.16161616 0.2020202
 0.24242424 0.28282828 0.32323232 0.36363636 0.4040404  0.44444444
 0.48484848 0.52525253 0.56565657 0.60606061 0.64646465 0.68686869
 0.72727273 0.76767677 0.80808081 0.84848485 0.88888889 0.92929293
 0.96969697 1.01010101 1.05050505 1.09090909 1.13131313 1.17171717
 1.21212121 1.25252525 1.29292929 1.33333333 1.37373737 1.41414141
 1.45454545 1.49494949 1.53535354 1.57575758 1.61616162 1.65656566
 1.6969697  1.73737374 1.77777778 1.81818182 1.85858586 1.8989899
 1.93939394 1.97979798 2.02020202 2.06060606 2.1010101  2.14141414
 2.18181818 2.22222222 2.26262626 2.3030303  2.34343434 2.38383838
 2.42424242 2.46464646 2.50505051 2.54545455 2.58585859 2.62626263
 2.66666667 2.70707071 2.74747475 2.78787879 2.82828283 2.86868687
 2.90909091 2.94949495 2.98989899 3.03030303 3.07070707 3.11111111
 3.15151515 3.19191919 3.23232323 3.27272727 3.31313131 3.35353535
 3.39393939 3.43434343 3.47474747 3.51515152 3.55555556 3.595959

## Esercizio 4
Costruire il numpy array 1d C  di valori che variano tra 15 e 41 con step 2

-     Trasformarlo in un numpy array 2d con shape(7,2) usando il metodo . resize(7,2)  che lavora in-place,  (cioè  modifica l'array esistente).  

In [14]:
c = np.arange(15, 42, 2)
c.resize(7, 2)
print(c)

[[15 17]
 [19 21]
 [23 25]
 [27 29]
 [31 33]
 [35 37]
 [39 41]]


 ## Esercizio 5
 Creare un array NumPy di 10 numeri interi casuali tra 1 e 100.
 

In [15]:
arr = np.random.randint(1, 101, 10)
print(arr)

[86 18 84 38  5 97 96 93  6 74]


 ## Esercizio 6
 Eseguire uno shuffle degli elementi di un numpy-array di 30 numeri interi casuali tra 100 e 200.


In [17]:
arr = np.random.randint(100, 201, 30)
np.random.shuffle(arr)
print(arr)

[193 162 189 108 120 124 154 116 150 174 141 197 145 136 179 118 122 196
 107 148 134 162 156 167 122 156 122 200 134 122]


  ## Esercizio 7
 Creare una permutazione casuale degli elementi di un numpy-array di 20 valori campionati da una distribuzione  gaussiana con media zero e varianza 1, senza modificarlo.

In [18]:
arr = np.random.randn(20)
arr = np.random.permutation(arr)
print(arr)

[-3.76636572e-01 -4.10300399e-01  4.37374761e-01 -1.39029360e-01
 -1.31198544e+00  6.02271825e-01 -7.83354249e-01 -8.96357861e-02
 -4.10567615e-01  2.31544402e-01 -1.00047555e+00 -1.87631309e-01
 -5.69938501e-01  1.73545430e+00 -6.76232183e-04 -1.70982718e+00
  1.34313502e+00  9.86620427e-01  1.29848392e-02  8.38827726e-01]


 ## Esercizio 8
-     Scrivere il numpy  array 2d, A, [[10,11,12],[13,14,15],[16,17,18]] 
-     trasformarlo in un numpy array 1d usando il metodo .ravel()  
-     Si tratta di una copia o di una vista?

In [43]:
a = np.array(range(10, 19)).reshape(3, 3)
b = a.ravel()
print(b)
print(b.base)

[10 11 12 13 14 15 16 17 18]
[10 11 12 13 14 15 16 17 18]


## Esercizio 9
-     Scrivere il numpy  array 2d, A, [[10,11,12],[13,14,15],[16,17,18]] 
-     trasformarlo in un numpy array 1d usando il metodo  .flatten().
-     Si tratta di una copia o di una vista?

In [42]:
a = np.array(range(10, 19)).reshape(3, 3)
b = a.flatten()
print(b)
print(b.base)

[10 11 12 13 14 15 16 17 18]
None


 ## Esercizio 10
-     creare i due  numpy array 1d
        f1= [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15]  
        f2= [16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31]
-     "vederli" come matrici  4x4 ed effettuarne il prodotto scalare, memorizzando in f3 la matrice che nasce dal loro prodotto.

In [29]:
f1 = np.array(range(16))
f2 = np.array(range(16, 32))
a = f1.view()
b = f2.view()
a.resize(4, 4)
b.resize(4, 4)
f3 = a @ b

print(f3)

[[ 152  158  164  170]
 [ 504  526  548  570]
 [ 856  894  932  970]
 [1208 1262 1316 1370]]


## Esercizio 11
Costruire:
 -     Numpy Array A di shape (2,10) : array di valori random campionati da una distribuzione gaussiana con media 0 e varianza 1.
 -     Numpy array 1D B di shape (20,) formato dagli interi da 0 a 19.
 
 "Vedere" A come un array 1D di 20 elementi ed effettuare la somma di A+B


In [41]:
a = np.random.rand(20).reshape(2, 10)
b = np.array(range(20))
a = a.ravel() + b

print(a)
print(a.base)

[ 0.65477449  1.15263871  2.35041857  3.84731846  4.74708757  5.17475554
  6.40969777  7.11717226  8.37875026  9.48764688 10.71963199 11.34255193
 12.17139282 13.71484712 14.76908991 15.1318351  16.1259768  17.10369479
 18.01824266 19.23566042]
None


## Esercizio 12
-     Costruire f, un numpy array 1D , di shape (10,) costituito da 10 valori equidistanti nell'intervallo [0,1]
-     aumentarne la dimensionalità, in maniera tale che abbia shape (10,1)

In [36]:
f = np.linspace(0, 1, 10).reshape(10, 1)

print(f)

[[0.        ]
 [0.11111111]
 [0.22222222]
 [0.33333333]
 [0.44444444]
 [0.55555556]
 [0.66666667]
 [0.77777778]
 [0.88888889]
 [1.        ]]


## Esercizio 13
-     Costruire i seguenti numpy array
      a= np.array([14,13,12,11])
      b=np.array([[4,3,2,1],[9,10,11,12]])

  Costruire il numpy array che si ottiene concatenando a e b per righe

In [57]:
a = np.array([14,13,12,11]).reshape(1, 4)
b = np.array([[4,3,2,1],[9,10,11,12]])
arr = np.concatenate((a, b), axis=0)

print(arr)

[[14 13 12 11]
 [ 4  3  2  1]
 [ 9 10 11 12]]


## Esercizio 14
-     Costruire i seguenti numpy array
      a= np.array([14,13,12,11])
      b=np.array([[4,3,2,1],[9,10,11,12],[4,3,2,1],[9,10,11,12]])

-
- Costruire il numpy array che si ottiene concatenando a e b per colonne

## Esercizio 15
Costruire, facendo uso di una sola riga di codice, la matrice A di ordine 10, che ha gli unici elementi diversi da zero, sulla diagonale principale, sulla codiagonale inferiore e sulla codiagonale superiore:

elementi diagonali $A_{i,i}=10 \quad$, i=1,20

elementi nella codiagonale inferiore $A_{i-1,i}=-12, \quad$ i=2,..,20, 

elementi nella codiagonale superiore $A_{i,i+1}=-2 \quad$ i=1,..,19

## Esercizio 16
Costruire la matrice identità, denominata *I*,  della stessa dimensione di *A*

## Esercizio 17
Facendo uso dell'operatore slicing *:* ,costruire la matrice AA selezionando dalla matrice A le righe che vanno dalla 2 alla 9 con step 2 e le colonne che vanno dalla 3 alla 6 con step 2.
La matrice AA è una vista o una copia?

## Esercizio 18
Facendo uso dell'operatore slicing *:*, costruire la matrice A1 selezionando dalla matrice A le colonne che vanno dalla 1 alla 5 con step 2.  Trasforma la matrice A1 in una copia.

## Esercizio 19

Costruisci due numpy array A e B di shape (5,5) 
   - A formata da numeri interi casuali  nell'intervallo [1,20]
   - B formata da numeri interi casuali  nell'intervallo [1,6]
 
  Costruisci con una sola riga di codice le seguenti matrici
  
  C=A+B   dove  $\quad c_{ij}=a_{ij}+b_{ij} \quad i=0,..,n, \quad j=0,...,n$
  
  D=A-B   dove  $\quad d_{ij}=a_{ij}-b_{ij} \quad i=0,..,n, \quad j=0,...,n$
  
  E=A*B   dove  $\quad e_{ij}=a_{ij}*b_{ij} \quad i=0,..,n, \quad j=0,...,n$
  
  F=A/B   dove  $\quad f_{ij}=a_{ij}/b_{ij} \quad i=0,..,n, \quad j=0,...,n$

## Esercizio 20
Dato array NumPy che rappresenta le vendite giornaliere di un negozio per un mese,
vendite = np.random.randint(50, 200, 30) 

        Trova tutti i giorni in cui le vendite sono state superiori a 150.

        Calcola la media delle vendite solo per i giorni in cui sono state superiori a 150.

        Imposta le vendite a 0 per tutti i giorni in cui sono state inferiori a 100.


## Esercizio 21
Costruisci A, un numpy array 2d di shape (3,3) e b,  un numpy array 2d di shape (3,1)
e somma b ad ogni colonna di A b


# Esercizio 22
dato il numpy array 2d G=np.linspace(0,10,20).reshape(-1,4), scambia in G la seconda riga con la quarta riga

## Esercizio 23
Visualizzare il grafico della funzione $f(x)=sin^2(x)/x$ in 200 punti equidistanti nell'intervallo $[0.1,2\pi]$

## Esercizio 24

Data il numpy arra2d ottenuto facendo un reshape a (4,4) di un numpy array 1d di 16 valori equidistanti nell'intervallo [0,4]
ed un numpy array 1d di 4 valori  interi casuali nell'intervallo [10,60], costruire il prodotto scalare A per b,
usando A@b, e vettorizzando solo il ciclo interno.


## Esercizio 25
Dato g un numpy array 2d, estendilo di una riga in alto ed di una riga in basso, di una colonna a sinistra ed di una colonna a destra, usa la funzione g_est= np.pad(g,(1,1),mode='reflect'). Visualizza 'g_est' e commenta cosa vedi.

## Esercizio 26
Scrivi un programma Python che esegua l'elaborazione di un'immagine utilizzando un filtro di media. 

Utilizza il padding per gestire i pixel del bordo durante l'applicazione del filtro.

Il programma deve seguire questi passaggi:

- Carica l'immagine: Carica l'immagine 'cameraman.png' utilizzando la libreria Matplotlib.**image=plt.imread('cameraman.png')**
- Visualizza l'immagine originale **plt.imshow(image,cmap='gray')**
- definisce la dimensione del filtro, dim_filtro
- Applica il padding: Applica un padding costante di dim_filtro//2 pixel a tutti i bordi dell'immagine (utilizzando il metodo di numpy **image_bordered=np.pad(image,(dim_filtro//2,dim_filtro//2),, mode='reflect')**
- Crea il filtro di media: Crea un filtro di media di dimensioni *dim_filtroxdim_filtro*, dove tutti gli elementi sono pari a 1/(dim_filtroxdim_filtro)
- Applica il filtro: Scorri ogni pixel dell'immagine con padding (escludendo i bordi aggiunti) e applica il filtro di media.

   *Per ogni pixel, calcola la somma dei prodotti tra i pixel della sottomatrice dim_filtroxdim_filtro centrata sul pixel corrente e il filtro.*
   
- Crea l'immagine filtrata: Crea un nuovo array NumPy che rappresenta l'immagine filtrata, dove ogni pixel contiene il risultato dell'applicazione del filtro.
- Visualizza l'immagine filtrata: Visualizza l'immagine filtrata utilizzando Matplotlib con la mappa di colori 'gray'.
- Aggiungi una colorbar per visualizzare la scala dei valori dell'immagine filtrata.
- Mostra l'immagine filtrata con Matplotlib.


Requisiti:

-  Utilizzare la libreria Matplotlib per caricare e visualizzare l'immagine.
-  Utilizzare la libreria NumPy per la manipolazione degli array e per il padding.